### ISSUE TITLE
fully qualified call required to determine trait method type

### ISSUE THREAD

**Issue by jimblandy** [rust-lang/rust-analyzer#12247]
Sometimes `rust-analyzer` cannot determine the type returned by calling a trait method, unless fully qualified call syntax is used. It may be that it cannot even determine which method is being called.

The branch `https://github.com/jimblandy/wgpu/tree/placate-rust-analyzer` shows an example: without the change in jimblandy/wgpu@77468299ca177aee2fd2005aed939bb28c7dd89b, `rust-analyzer` cannot determine the type of `cmd_buf`. Switching to fully-qualified syntax and providing some trait type parameters allows `rust-analyzer` to see that `cmd_buf` should have type `&mut CommandBuffer<A>`.

Here's the change:

```
modified   wgpu-core/src/command/compute.rs
@@ -309,8 +309,9 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         let (device_guard, mut token) = hub.devices.read(&mut token);
 
         let (mut cmd_buf_guard, mut token) = hub.command_buffers.write(&mut token);
-        let cmd_buf = CommandBuffer::get_encoder_mut(&mut *cmd_buf_guard, encoder_id)
-            .map_pass_err(init_scope)?;
+        let result = CommandBuffer::get_encoder_mut(&mut *cmd_buf_guard, encoder_id);
+        // Call using fully qualified syntax, to placate rust-analyzer.
+        let cmd_buf = MapPassErr::<&mut CommandBuffer<A>, _>::map_pass_err(result, init_scope)?;
         // will be reset to true if recording is done without errors
         cmd_buf.status = CommandEncoderStatus::Error;
         let raw = cmd_buf.encoder.open();
```

**Comment by jimblandy** [comment]
Here's another fix that works:

```
modified   wgpu-core/src/command/compute.rs
@@ -309,7 +309,9 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         let (device_guard, mut token) = hub.devices.read(&mut token);
 
         let (mut cmd_buf_guard, mut token) = hub.command_buffers.write(&mut token);
-        let cmd_buf = CommandBuffer::get_encoder_mut(&mut *cmd_buf_guard, encoder_id)
+        // Spell out the type, to placate rust-analyzer.
+        // https://github.com/rust-lang/rust-analyzer/issues/12247
+        let cmd_buf: &mut CommandBuffer<A> = CommandBuffer::get_encoder_mut(&mut *cmd_buf_guard, encoder_id)
             .map_pass_err(init_scope)?;
         // will be reset to true if recording is done without errors
         cmd_buf.status = CommandEncoderStatus::Error;
```

**Comment by jimblandy** [comment]
In case GitHub doesn't keep that other commit around, here's a commit that made it to `master` where, before the commit, rust-analyzer can't determine `cmd_buf`'s type, but after the commit, it can:

gfx-rs/wgpu@284ed4654e6bd79a4d883b766cb70a9474c06a01

**Comment by HKalbasi** [comment]
I believe this will be fixed once #14279 becomes fixed.
